tests/components/humidifier/test_reproduce_state.py

```diff
@@ -0,0 +1,237 @@
+    assert len(humidity_calls) == 2
+    # order is not guaranteed
+    assert any(
+        call.data == {"entity_id": ENTITY_1, "humidity": 40} for call in humidity_calls
+    )
+    assert any(
+        call.data == {"entity_id": ENTITY_2, "humidity": 50} for call in humidity_calls
+    )
+
+
+async def test_state_with_none(hass):
+    """Test that none is not a humidifier state."""
+    hass.states.async_set(ENTITY_1, STATE_OFF, {})
+
+    turn_on_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_ON)
+    turn_off_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_OFF)
+    mode_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_MODE)
+    humidity_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_HUMIDITY)
+
+    await async_reproduce_states(hass, [State(ENTITY_1, None)])
+
+    await hass.async_block_till_done()
+
+    assert len(turn_on_calls) == 0
+    assert len(turn_off_calls) == 0
+    assert len(mode_calls) == 0
+    assert len(humidity_calls) == 0
+
+
+async def test_state_with_context(hass):
+    """Test that context is forwarded."""
+    hass.states.async_set(ENTITY_1, STATE_OFF, {})
+
+    turn_on_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_ON)
+    turn_off_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_OFF)
+    mode_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_MODE)
+    humidity_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_HUMIDITY)
+
+    context = Context()
+
+    await async_reproduce_states(
+        hass,
+        [State(ENTITY_1, STATE_ON, {ATTR_MODE: MODE_AWAY, ATTR_HUMIDITY: 45})],
+        context=context,
+    )
+
+    await hass.async_block_till_done()
+
+    assert len(turn_on_calls) == 1
+    assert turn_on_calls[0].data == {"entity_id": ENTITY_1}
+    assert turn_on_calls[0].context == context
+    assert len(turn_off_calls) == 0
+    assert len(mode_calls) == 1
+    assert mode_calls[0].data == {"entity_id": ENTITY_1, "mode": "away"}
+    assert mode_calls[0].context == context
+    assert len(humidity_calls) == 1
+    assert humidity_calls[0].data == {"entity_id": ENTITY_1, "humidity": 45}
+    assert humidity_calls[0].context == context
+
+
+@pytest.mark.parametrize(
+    "service,attribute",
+    [(SERVICE_SET_MODE, ATTR_MODE), (SERVICE_SET_HUMIDITY, ATTR_HUMIDITY)],
+)
+async def test_attribute(hass, service, attribute):
+    """Test that service call is made for each attribute."""
+    hass.states.async_set(ENTITY_1, STATE_ON, {})
+
+    turn_on_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_ON)
+    turn_off_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_OFF)
+    calls_1 = async_mock_service(hass, DOMAIN, service)
+
+    value = "dummy"
+
+    await async_reproduce_states(hass, [State(ENTITY_1, STATE_ON, {attribute: value})])
+
+    await hass.async_block_till_done()
+
+    assert len(turn_on_calls) == 0
+    assert len(turn_off_calls) == 0
+    assert len(calls_1) == 1
+    assert calls_1[0].data == {"entity_id": ENTITY_1, attribute: value}
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5745133)